### DISCUSSION TITLE
[경훈] assignment 기능구현

### DISCUSSION THREAD

**Discussion by tirhande** [Wanted-07-team-9/pre-onboarding-7th-3-1-9#3] [기능구현]
### 0️⃣ json-server 배포
- localhost가 아닌 vercel에 json-server를 배포하여 사용
- ex) https://json-server-beryl.vercel.app/api/sick?sickNm_like=간
- ex) https://json-server-beryl.vercel.app/api/sick?q=콜레라

## **💡 Assignment 목차**
- [1️⃣ Assignment](#1️⃣-assignment)
- [2️⃣ Assignment](#2️⃣-assignment)
- [3️⃣ Assignment](#3️⃣-assignment)
- [4️⃣ Assignment](#4️⃣-assignment)
- [5️⃣ Assignment](#5️⃣-assignment)
- [6️⃣ Assignment](#6️⃣-assignment)
#
### **1️⃣ Assignment**

- 질환명 검색시 API 호출 통해서 검색어 추천 기능 구현
  - 사용자가 입력한 텍스트와 일치하는 부분 볼드처리

⭐️⭐️⭐️⭐️⭐️
### array를 map으로 반복하면서 split된 부분마다 "검색한 단어"를 삽입하여 bold 처리 하였습니다.
⭐️⭐️⭐️⭐️⭐️

<p align="center">
<img width="537" alt="search" src="https://user-images.githubusercontent.com/74575497/200851310-d8d964fc-5e51-4b17-aca4-211d7f09c8cc.png">
</p>

https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/bbf1724458dd765687641bed9246447632d70bc6/src/components/blocks/AutoCompleteItem.tsx#L4-L28

# 
### **2️⃣ Assignment**

- - 검색어가 없을 시 “검색어 없음” 표출

⭐️⭐️⭐️⭐️⭐️
### 검색된 문자열이 없을경우 검색어 없음 표출하였습니다.
⭐️⭐️⭐️⭐️⭐️

<p align="center">
<img width="528" alt="empty" src="https://user-images.githubusercontent.com/74575497/200851832-41b84cc0-e62e-4d6d-b2ea-9515c94069bc.png">
</p>

https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/bbf1724458dd765687641bed9246447632d70bc6/src/components/blocks/EmptyResult.tsx#L3-L9
https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/bbf1724458dd765687641bed9246447632d70bc6/src/pages/Main/AutoComplete/AutoComplete.tsx#L35

# 
### **3️⃣ Assignment**

- API 호출 최적화
  - API 호출별로 로컬 캐싱 구현
    - 캐싱 기능을 제공하는 라이브러리 사용 금지(React-Query 등)
    - 캐싱을 어떻게 기술했는지에 대한 내용 README에 기술

⭐️⭐️⭐️⭐️⭐️
### Cache Storage에 없을경우 API 호출 진행
⭐️⭐️⭐️⭐️⭐️

<p align="center">
<img width="420" alt="cache" src="https://user-images.githubusercontent.com/74575497/200842174-e1b72327-5569-4b5e-b7d7-594ec5499587.png">
</p>

https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/bbf1724458dd765687641bed9246447632d70bc6/src/apis/ClinicalService.ts#L4-L31
# 
### **4️⃣ Assignment**

- - 입력마다 API 호출하지 않도록 API 호출 횟수를 줄이는 전략 수립 및 실행
    - README에 전략에 대한 설명 기술

⭐️⭐️⭐️⭐️⭐️
### 사용자의 입력마다 API 호출 하는것이 아닌 사용자의 입력이 해당 시간동안 없을때만 호출하게 하여 횟수를 줄였습니다.
⭐️⭐️⭐️⭐️⭐️

https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/bbf1724458dd765687641bed9246447632d70bc6/src/hooks/useGetSickItem.ts#L16-L29

# 
### **5️⃣ Assignment**

- - API를 호출할 때 마다 `console.info("calling api")` 출력을 통해 콘솔창에서 API 호출 횟수 확인이 가능하도록 설정

⭐️⭐️⭐️⭐️⭐️
### axios instance의 인터셉터에서 request 전에 console.info('calling api'); 를 추가했습니다.
⭐️⭐️⭐️⭐️⭐️

<p align="center">
<img width="384" alt="console" src="https://user-images.githubusercontent.com/74575497/200847063-91f2c08f-9871-4c3c-bac6-6a2dd94a48ba.png">
</p>

https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/bbf1724458dd765687641bed9246447632d70bc6/src/apis/axiosInstance.ts#L8-L16

# 
### **6️⃣ Assignment**

- 키보드만으로 추천 검색어들로 이동 가능하도록 구현
  - 사용법 README에 기술

⭐️⭐️⭐️⭐️⭐️
### selectIndex값과 아이템의 index 값이 같으면 over className을 추가하였습니다.
⭐️⭐️⭐️⭐️⭐️

https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/bbf1724458dd765687641bed9246447632d70bc6/src/pages/Main/SearchBar/index.tsx#L14-L23
https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/bbf1724458dd765687641bed9246447632d70bc6/src/components/blocks/AutoCompleteItem.tsx#L16-L26

**Comment by jun-05** [comment]
캐시 스토리지 기능을 잘 이해하고 사용하신것 같습니다. 
저같이 매 캐쉬스토리지를 오픈하는것보다, 하나에 쌓아두는 방식이 더 옳은 방식이라고 생각했습니다.
키보드 이벤트가 제일 윗쪽 값보다 한칸 더 가면 ui가 제대로 작동하지 않는게 아쉬웠습니다
